### PR TITLE
Allow casting of bools in string representation to to true bools

### LINKED ISSUE #57256: RFC: Automatic cast of bool literals in obvious cases
```sql
create table tab (d Date, b bool) engine = Memory;

insert into tab values ('2023-11-27', 'true'); -- insert of string-encoded bool works

select d = '2023-11-27' from tab;

select b = true from tab;
select b = 'true' from tab; -- throws: comparison with string-encoded bool does not work
```

Booleans should at least be able to compare against obvious string literals such as `true'`, `'1'`, `'enabled'` etc.

Related issue: #11630

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #57256

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Boolean values in string representation now cast to true bools. E.g. this query previously threw an exception but now works: `SELECT true = 'true'`.

### CODE REVIEW

File: tests/queries/0_stateless/02933_compare_with_bool_as_string.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+select true = 'true';
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02933_compare_with_bool_as_string.sql:1]: This PR has twice as many test lines as changed code lines.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 6b6630c5a7cc0ed49eb1ef22b139684b3ba22c20 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60160/6b6630c5a7cc0ed49eb1ef22b139684b3ba22c20/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60160/6b6630c5a7cc0ed49eb1ef22b139684b3ba22c20/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by nikitamikhaylov** [comment]
Are we in JavaScript? Do we need to introduce somewhat like `===` which won't perform type conversion?

**Comment by rschu1ze** [comment]
Didn't know about JS's `===` 😄  but I don't think we need this in SQL.

This PR only affects domain data types (e.g. `bool`). Right now, the wrong serialization is chosen which leads to surprises.

**Review by SmitaRKulkarni — APPROVED** [review]
LGTM

**Pull request merged**